sys/net/network_layer/ng_sixlowpan/frag/rbuf.c

```diff
@@ -0,0 +1,307 @@
+#include "net/ng_sixlowpan.h"
+#include "net/ng_sixlowpan/frag.h"
+#include "thread.h"
+#include "timex.h"
+#include "vtimer.h"
+#include "utlist.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define RBUF_INT_SIZE   (11 * RBUF_SIZE)    /* 11 == default MTU / 127 */
+
+static rbuf_int_t rbuf_int[RBUF_INT_SIZE];
+
+static rbuf_t rbuf[RBUF_SIZE];
+
+#if ENABLE_DEBUG
+static char l2addr_str[3 * RBUF_L2ADDR_MAX_LEN];
+#endif
+
+/* ------------------------------------
+ * internal function definitions
+ * ------------------------------------*/
+/* checks whether start and end are in given interval i */
+static inline bool _rbuf_int_in(rbuf_int_t *i, uint16_t start, uint16_t end);
+/* gets a free entry from interval buffer */
+static rbuf_int_t *_rbuf_int_get_free(void);
+/* remove entry from reassembly buffer */
+static void _rbuf_rem(rbuf_t *entry);
+/* update interval buffer of entry */
+static bool _rbuf_update_ints(rbuf_t *entry, uint16_t offset, size_t frag_size);
+/* checks timeouts and removes entries if necessary (oldest if full) */
+static void _rbuf_gc(void);
+/* gets an entry identified by its tupel */
+static rbuf_t *_rbuf_get(const void *src, size_t src_len,
+                         const void *dst, size_t dst_len,
+                         size_t size, uint16_t tag);
+
+void rbuf_add(ng_pktsnip_t *pkt)
+{
+    ng_netif_hdr_t *hdr = pkt->next->data;
+    rbuf_t *entry;
+    ng_sixlowpan_frag_t *frag = pkt->next->data;
+    size_t datagram_size;
+    size_t frag_size;
+    uint16_t tag, offset = 0;
+    kernel_pid_t iface = hdr->if_pid;
+    uint8_t *data;
+    bool first;
+
+    switch (frag->disp_size.u8[0] & NG_SIXLOWPAN_FRAG_DISP_MASK) {
+        case NG_SIXLOWPAN_FRAG_1_DISP:
+            tag = byteorder_ntohs(frag->tag);
+            datagram_size = byteorder_ntohs(frag->disp_size) & NG_SIXLOWPAN_FRAG_SIZE_MASK;
+            frag_size = (pkt->size - sizeof(ng_sixlowpan_frag_t));
+            data = (uint8_t *)(hdr + 1);
+            first = true;
+            break;
+
+        case NG_SIXLOWPAN_FRAG_N_DISP:
+            tag = byteorder_ntohs(frag->tag);
+            datagram_size = byteorder_ntohs(frag->disp_size) & NG_SIXLOWPAN_FRAG_SIZE_MASK;
+            offset = (((ng_sixlowpan_frag_n_t *)frag)->offset * 8);
+            frag_size = (pkt->size - sizeof(ng_sixlowpan_frag_n_t));
+            data = (uint8_t *)(hdr + 1) + 1;
+            break;
+
+        default:
+            DEBUG("6lo rbuf: Not a fragment header.\n");
+            ng_pktbuf_release(pkt);
+
+            return;
+    }
+
+    _rbuf_gc();
+    entry = _rbuf_get(ng_netif_hdr_get_src_addr(hdr), hdr->src_l2addr_len,
+                      ng_netif_hdr_get_dst_addr(hdr), hdr->dst_l2addr_len,
+                      datagram_size, tag);
+
+    if (entry == NULL) {
+        DEBUG("6lo rbuf: reassembly buffer full.\n");
+        ng_pktbuf_release(pkt);
+
+        return;
+    }
+
+    if (_rbuf_update_ints(entry, offset, frag_size)) {
+        if (first) {
```

Comment:
No. There is code that must be executed if `first` and if `!first` below the second if.